[Backend receipts] Error handling

iamgabrielma opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes #11829

## Description
This PR adds error handling by logging the error internally as well as displaying a user-facing notification for the cases when this could fail, that is:
- We fail to fetch a receipt via tapping on `See receipt``
- We fail to print a receipt via "print or share" receipt when performing a card present payment.

| See Receipt failed | Print receipt failed |
|--------|--------|
| <img width=300 src="https://github.com/woocommerce/woocommerce-ios/assets/3812076/26f61358-670d-4d67-aef1-a342844ec833"> | <img width="302" alt="Screenshot 2024-02-06 at 16 22 10" src="https://github.com/woocommerce/woocommerce-ios/assets/3812076/fe6f35cc-7e90-4fbf-9c2b-7286a6beff0d"> | 

## Testing instructions
Return early with an error in `ReceiptStore.retrieveReceipt(:)` method, 
```diff
func retrieveReceipt(order: Order, onCompletion: @escaping (Result<Receipt, Error>) -> Void) {
+ onCompletion(.failure(NSError(domain: "oopsie", code: 123)))
+ return
remote.retrieveReceipt(siteID: order.siteID,
... 
```

As well as in `PaymentCaptureOrchestrator.presentBackendReceipt(:)`
```diff
func presentBackendReceipt(for order: Order, onCompletion: @escaping (Result<Receipt, Error>) -> Void) {
+        onCompletion(.failure(NSError(domain: "oopsie", code: 123)))
+        return
```
This way we can be sure that we're getting an error in both flows. Now, on a eligible store (woo-dev plugin with receipts API, you can use [indiemelon.mystagingwebsite.com](indiemelon.mystagingwebsite.com), since this one has WCPay enabled for testing):
1. Proceed to create an order. Collect cash payment. Tap on "See receipt". Observe that we fail to fetch the receipt and the notice appears in the bottom of the view.
2. Proceed to create a new order. Collect payment via card present. Tap on "Print receipt". Observe the notice appears in the bottom of the view.

### Caveats
If the 2nd flow goes fast enough, the user won't see the notice since we navigate them automatically to the order details upon completion, I don't think is worth to alter the whole navigation so they can see the notice, specially since the "See receipt" button will appear immediately after as well so they can try again. We can attempt to improve this in the future, but for the moment, if you need to slow it down for testing, we can delay the completion in the `presentBackedReceiptFailedNotice()` call as such:
```diff
-onCompleted()
+DispatchQueue.main.asyncAfter(deadline: .now() + 3) {
+	onCompleted()
+}
```
![Simulator Screen Recording - iPhone 15 - 2024-02-06 at 16 05 24](https://github.com/woocommerce/woocommerce-ios/assets/3812076/e955b4da-4655-4efc-bbf7-ba20c034aa07)

iamgabrielma committed:
Handle error on order detail receipt retrieval

iamgabrielma committed:
Update presentBackendReceipt with Result type

iamgabrielma committed:
Handle error when fail to print card present receipt

dangermattic commented:
<!--
  1 Error: PR requires at least one label...
  1 Warning: PR is not assigned to a milest...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">PR requires at least one label.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">PR is not assigned to a milestone.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

iamgabrielma committed:
Update wcPluginDevVersion

Due to the API ongoing development, we have multiple sites with different wc dev versions. In order to avoid problems when debugging, I’ve added all of them to the eligibility check. We’ll remove this as soon as the API is merged into core trunk.

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/185">17.3</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

bozidarsevo commented on WooCommerce/Classes/ViewModels/Order Details/Receipts/ReceiptEligibilityUseCase.swift:
```diff
@@ -44,6 +44,10 @@ private extension ReceiptEligibilityUseCase {
         static let wcPluginMinimumVersion = "8.7.0"
-        static let wcPluginDevVersion: [String] = ["8.6.0-dev-7625495467-gf50cc6b", "8.6.0-dev-7708067547-g87e8c5f"]
+        static let wcPluginDevVersion: [String] = ["8.7.0-dev",
+                                                   "8.6.0-dev",
+                                                   "8.5.0-beta.1",
+                                                   "8.6.0-dev-7625495467-gf50cc6b",
+                                                   "8.6.0-dev-7708067547-g87e8c5f"]
```
Just for me to learn, "8.6.0-dev" is not enough to cover all 8.6.0 versions, we need the other explicit ones like "8.6.0-dev-7625495467-gf50cc6b" also?

iamgabrielma replied:
That's right, the plugin version can be seen if you go to the plugin's main PHP file, where the header is. eg: in `woocommerce.php` for one of the experimental branches you'll see: 
<img width="557" alt="Screenshot 2024-02-07 at 21 48 49" src="https://github.com/woocommerce/woocommerce-ios/assets/3812076/daf688df-f3ed-479d-97ad-a19677875de3">

So when we get the `SystemPlugin` from the remote, we know from one of the properties we fetch what's the exact version of the installed plugin. 

We have helper tools in `VersionHelpers.swift` to compare, for example, `minimumVersion`, so let's say that any version 8.6.0+ would be valid if that's what we pass in. We _might_ have also methods to compare dev versions from what I see [here](https://github.com/woocommerce/woocommerce-ios/blob/trunk/WooCommerce/Classes/Tools/VersionHelpers.swift#L117), but to be honest I haven't tested those 😅
